compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/types/irTypes.kt

```diff
@@ -41,14 +45,40 @@ val IrType.classifierOrNull: IrClassifierSymbol?
 val IrType.classOrNull: IrClassSymbol?
     get() = classifierOrNull as? IrClassSymbol
 
-fun IrType.makeNotNull() =
-    if (this is IrSimpleType && this.hasQuestionMark) {
-        buildSimpleType {
-            kotlinType = originalKotlinType?.makeNotNullable()
-            hasQuestionMark = false
+val IrType.isNullable: Boolean
+    get() = isMarkedNullable() || this is IrDynamicType ||
+            classifierOrNull?.safeAs<IrTypeParameterSymbol>()?.owner?.superTypes?.all { it.isNullable } == true
+
+fun IrType.makeNotNull(): IrType {
+    if (!this.isNullable || this !is IrSimpleType) {
+        // This is wrong for IrDynamicType, but there's nothing we can do about it.
+        return this
+    }
+
+    val newClassifier = if (classifier is IrTypeParameterSymbol) {
+        val descriptor = WrappedTypeParameterDescriptor()
+        val symbol = IrTypeParameterSymbolImpl(descriptor)
+        val newParameter = with(classifier.owner as IrTypeParameter) {
+            IrTypeParameterImpl(startOffset, endOffset, origin, symbol, name, index, isReified, variance).also {
```

Comment:
I'm not sure if this is the right approach. I'd prefer just always creating some sort of `IrNotNullTypeImpl` which is a subtype of `IrSimpleType` (or `IrSimpleTypeImpl`) and returns `true` on `isNullable`. Similarly to how it's done for `KotlinType` implementations (see [`KotlinTypeFactory.kt`](https://github.com/JetBrains/kotlin/blob/master/core/descriptors/src/org/jetbrains/kotlin/types/KotlinTypeFactory.kt)). Creating fake IR (with fake symbols) is the one thing that we want to avoid where it's possible.

Reply:
Yes, that is a better approach and also fixes an issue in psi2ir (there's a workaround in the PR). I didn't consider it, because I didn't want to change the IR, but if you agree I would like to change the implementation in this manner.

Reply:
@dnpetrov @romanart 
1) Do you have any objections?
2) Would this conflict with the ongoing branch on IrType subtyping very much? I suppose it'd make sense to wait until that branch is merged in that case

Reply:
I don't mind conflicts as soon as we have corresponding psi2ir tests.